### PR TITLE
fix(providers): broken logging

### PR DESCRIPTION
**Pull request by vados-cosmonic**
## Feature or Problem

This commit fixes the issue with logging in providers which was caused by our use of `set_global_default`[0] from within a library (in this case `wasmcloud-tracing`).

`set_global_default` should not be called from libraries, so the `configure_observability()` function needed to be modified to ensure that configured subscribers were passed *back out* (as `tracing::Dispatch`s), and usable from the calling code in a binary context.

[0]: https://docs.rs/tracing-core/0.1.32/tracing_core/dispatcher/fn.set_global_default.html

Resolves https://github.com/wasmCloud/wasmCloud/issues/2368

### CODE REVIEW

**Comment by vados-cosmonic** [comment]
NOTE: PR is failing tests due to #2379 -- once that merges and this PR is rebased we should be good to go -- please feel free to review the code that is here (not all providers are updated anyway, but the changes will look identical).

File: crates/provider-keyvalue-redis/src/lib.rs [hunk 1]
```diff
@@ -56,10 +58,38 @@ pub async fn run() -> anyhow::Result<()> {
...
+    pub fn name() -> &'static str {
+        "keyvalue-redis-provider"
+    }
+
     pub async fn run() -> anyhow::Result<()> {
-        let HostData { config, .. } = load_host_data().context("failed to load host data")?;
+        let HostData {
+            config,
+            otel_config,
+            structured_logging,
```
Comment (brooksmtownsend) [inline crates/provider-keyvalue-redis/src/lib.rs:69]: Do these changes need to be made on each provider that wants to initialize tracing/logging?

**Review by brooksmtownsend — COMMENTED** [review]
The change to the SDK looks good, just wondering if we need to modify _all_ providers or if we could make this a simpler change in the provider SDK (or if this was just specific to the kvredis provider)

File: crates/provider-keyvalue-redis/src/lib.rs (hunk 1 above)
Comment (vados-cosmonic) [reply crates/provider-keyvalue-redis/src/lib.rs:69]: Yup, they do -- See this bit in the PR/commit:

> set_global_default should not be called from libraries, so the configure_observability() function needed to be modified to ensure that configured subscribers were passed back out (as tracing::Dispatchs), and usable from the calling code in a binary context.

Everything seems to be pointing to the fact that we should have never been calling `set_global_default` from inside the SDK/lib side to begin with -- they were interfering with each other.

All the other providers need to do this as well

Comment (vados-cosmonic) [reply crates/provider-keyvalue-redis/src/lib.rs:69]: As I [noted here](https://github.com/wasmCloud/wasmCloud/pull/2378#issuecomment-2186777538):

> NOTE: PR is failing tests due to https://github.com/wasmCloud/wasmCloud/pull/2379 -- once that merges and this PR is rebased we should be good to go -- please feel free to review the code that is here (not all providers are updated anyway, but the changes will look identical).

If the direction looks good, I can make this change to all the providers fairly quickly!

Comment (thomastaylor312) [reply crates/provider-keyvalue-redis/src/lib.rs:69]: Can we have a helper function in the sdk that handles init of logging?

Comment (vados-cosmonic) [reply crates/provider-keyvalue-redis/src/lib.rs:69]: Ah so we can't have it be a function but we *could* have a macro (that must get called from binary code somewhere). The problem here is that we can't... call the function that actually completes the setup from the SDK (as it's gonna be in multiple places/that's the problem)!

Maybe something like `wasmcloud_provider_sdk::init_tracing!("name-of-your-provider")` ?

Comment (vados-cosmonic) [reply crates/provider-keyvalue-redis/src/lib.rs:69]: basically, the `dispatch.try_init()` needs to *not* be called from library code. Everything else is fair game -- but the provider needs to hold on to the guard anyway as well for the duration of when it's running.

Comment (brooksmtownsend) [reply crates/provider-keyvalue-redis/src/lib.rs:69]: Oh yeah I'd be cool with a macro 👍🏻 Just something that simplifies the couple of extra steps, as long as the functions you have to call are documented that'll work fine

Comment (vados-cosmonic) [reply crates/provider-keyvalue-redis/src/lib.rs]: Done! Macro has been added, and all providers updated

File: crates/provider-keyvalue-redis/src/lib.rs [hunk 2]
```diff
@@ -56,10 +57,20 @@ pub async fn run() -> anyhow::Result<()> {
...
 
 impl KvRedisProvider {
+    pub fn name() -> &'static str {
+        "keyvalue-redis-provider"
+    }
+
     pub async fn run() -> anyhow::Result<()> {
+        initialize_observability!(
+            KvRedisProvider::name(),
+            std::env::var_os("PROVIDER_KV_REDIS_FLAMEGRAPH_PATH")
```
Comment (brooksmtownsend) [inline crates/provider-keyvalue-redis/src/lib.rs:67]: Do we actually pass this environment variable down to the provider?

**Review by brooksmtownsend — APPROVED** [review]
This looks _really_ nice with the macro 😄 I think it makes a lot of sense for our docs to note how to create the `init` function and then call this macro. I left a note for us to consider on the wasmCloud side but it doesn't necessitate blocking broken logging

File: crates/provider-keyvalue-redis/src/lib.rs (hunk 2 above)
Comment (vados-cosmonic) [reply crates/provider-keyvalue-redis/src/lib.rs:67]: Yeah this is for the case someone is running the provider locally (and trying to debug it), less so for when it's running in the host just normally. 

That said, we should probably have a way to pass more than a few env vars down to the child processes for providers!

Comment (vados-cosmonic) [reply crates/provider-keyvalue-redis/src/lib.rs:67]: Hmnnn maybe it's better to get that value from `HostData` *and* ENV (and use whichever is specified). Having an example of how to do that or a utility function might be quite nice.

**Pull request merged**